core/transaction.go

```diff
@@ -669,12 +669,16 @@ func TransactionCommitmentPoseidon(transactions []Transaction) (*felt.Felt, erro
 	var commitment *felt.Felt
 	return commitment, trie.RunOnTempTriePoseidon(commitmentTrieHeight, func(trie *trie.Trie) error {
 		for i, transaction := range transactions {
-			hashElems := []*felt.Felt{transaction.Hash()}
-			// todo handle empty signature
-			hashElems = append(hashElems, transaction.Signature()...)
-			hash := crypto.PoseidonArray(hashElems...)
+			var digest crypto.PoseidonDigest
+			digest.Update(transaction.Hash())
 
-			if _, err := trie.Put(new(felt.Felt).SetUint64(uint64(i)), hash); err != nil {
+			if txSignature := transaction.Signature(); len(txSignature) > 0 {
+				digest.Update(txSignature...)
+			} else {
+				digest.Update(&felt.Zero)
```

Comment:
"Each leaf in the patricia tree is Poseidon(transaction_hash, transaction_signature)"

Why do we hash zero if no signature was provided? I guess it's for historical reasons?

Reply:
This is the new commitment scheme, not all transaction have signatures such as Deploy and L1TransactionHandler. Therefore,  we have to add 0 felt see, here: https://github.com/starkware-libs/starknet-api/blob/15219992b54b170dcf80fa5cceab56847ea2f7f0/src/block_hash/transaction_commitment.rs#L32